From 4afde5aa827228d53c1c7df5a7999ff9394cb2ae Mon Sep 17 00:00:00 2001
From: Girija P <gpandiya@qti.qualcomm.com>
Date: Sun, 31 Aug 2025 23:51:29 +0530
Subject: [PATCH] Enabling MLO handling in NLHAL Fronthaul manager launch with
 vap & link id as identifier Backward compatible with legacy modes

Signed-off-by: Girija P <gpandiya@qti.qualcomm.com>
---
 .../fronthaul_manager/monitor/monitor.cpp     |  38 +-
 .../beerocks/slave/beerocks_agent_main.cpp    |  75 +++-
 .../bcl/include/bcl/beerocks_defines.h        |   6 +
 common/beerocks/bwl/common/base_wlan_hal.cpp  |  30 ++
 .../beerocks/bwl/include/bwl/base_wlan_hal.h  |   6 +
 .../beerocks/bwl/include/bwl/mon_wlan_hal.h   |   2 +
 .../{shared => include/bwl}/netlink_socket.h  |   0
 .../beerocks/bwl/include/bwl/nl80211_client.h |   6 +
 .../bwl}/nl80211_client_impl.h                |   8 +-
 .../{shared => include/bwl}/nl80211_socket.h  |   2 +-
 .../{shared => include/bwl}/nl_genl_socket.h  |   2 +-
 .../bwl/nl80211/ap_wlan_hal_nl80211.cpp       |  85 +++-
 .../bwl/nl80211/ap_wlan_hal_nl80211.h         |   4 +-
 .../bwl/nl80211/base_wlan_hal_nl80211.cpp     |  18 +-
 .../bwl/nl80211/mon_wlan_hal_nl80211.cpp      |  25 +-
 .../bwl/nl80211/mon_wlan_hal_nl80211.h        |   2 +
 common/beerocks/bwl/shared/netlink_socket.cpp |   2 +-
 .../bwl/shared/nl80211_client_factory.cpp     |   2 +-
 .../bwl/shared/nl80211_client_impl.cpp        | 371 ++++++++++++++----
 common/beerocks/bwl/shared/nl80211_socket.cpp |   2 +-
 common/beerocks/bwl/shared/nl_genl_socket.cpp |   2 +-
 .../platform/bpl/cfg/uci/bpl_cfg_wifi.cpp     |   2 +
 22 files changed, 555 insertions(+), 135 deletions(-)
 rename common/beerocks/bwl/{shared => include/bwl}/netlink_socket.h (100%)
 rename common/beerocks/bwl/{shared => include/bwl}/nl80211_client_impl.h (94%)
 rename common/beerocks/bwl/{shared => include/bwl}/nl80211_socket.h (98%)
 rename common/beerocks/bwl/{shared => include/bwl}/nl_genl_socket.h (97%)

diff --git a/agent/src/beerocks/fronthaul_manager/monitor/monitor.cpp b/agent/src/beerocks/fronthaul_manager/monitor/monitor.cpp
index 92078454b..a08117c74 100644
--- a/agent/src/beerocks/fronthaul_manager/monitor/monitor.cpp
+++ b/agent/src/beerocks/fronthaul_manager/monitor/monitor.cpp
@@ -15,6 +15,10 @@
 #include <bcl/network/sockets.h>
 #include <bcl/transaction.h>
 
+#ifdef USE_PRPLMESH_NLHAL
+#include <bwl/nl80211_client_impl.h>
+#endif
+
 #define BEEROCKS_CUSTOM_LOGGER_ID BEEROCKS_MONITOR
 #include <bcl/beerocks_logging_custom.h>
 #include <bpl/bpl_cfg.h>
@@ -224,7 +228,37 @@ bool Monitor::thread_init()
     }
     mon_db.set_clients_unicast_measurements(clients_unicast_measurements);
 
-    LOG(DEBUG) << "started";
+#ifdef USE_PRPLMESH_NLHAL
+    bwl::HALState state = mon_wlan_hal->get_state();
+    std::string interface_mld_mac;
+    state = mon_wlan_hal->get_state();
+
+    LOG(INFO) << "State of mon_wlan_hal is: " << static_cast<int>(state);
+    LOG(INFO) << "Interface OF monitor_iface is: " << monitor_iface;
+
+    size_t pos = monitor_iface.find(IFACE_LINK_SUFFIX);
+    if (pos != std::string::npos && pos > 0) {
+        std::string suffix = monitor_iface.substr(pos + IFACE_LINK_SUFFIX_LEN);
+        if (suffix.length() == 1 && std::isdigit(suffix[0])) {
+            int link_id = suffix[0] - '0';
+            if (link_id >= 0 && link_id <= 2) {
+                if(!mon_wlan_hal->get_mld_interface_mac(monitor_iface, interface_mld_mac)) {
+                    LOG(ERROR) << "get_mld_interface_mac call FAILED";
+                }
+                LOG(INFO) << "MLD MAC retrieved from nl hal: " << interface_mld_mac;
+
+                //Assign the mac address recived
+                m_radio_mac = tlvf::mac_from_string(interface_mld_mac);
+            } else {
+                LOG(WARNING) << "Invalid radio ID in interface name: " << monitor_iface;
+            }
+        } else {
+            LOG(WARNING) << "Invalid suffix format in interface name: " << monitor_iface;
+        }
+    } else {
+        LOG(WARNING) << "Interface name '" << monitor_iface << "' is not in expected format. Skipping MLD MAC retrieval.";
+    }
+#endif
 
     return true;
 }
@@ -689,7 +723,7 @@ bool Monitor::monitor_fsm()
                                 if(mapped_vap_element.num_links != -1) {
                                     int8_t num_links = mapped_vap_element.num_links;
                                     for (int it = 0; it < num_links; it++) {
-                                        std::string link_sock = mapped_vap_element.bss + "_link" + std::to_string(it);
+                                        std::string link_sock = mapped_vap_element.bss + IFACE_LINK_SUFFIX + std::to_string(it);
                                         mon_wlan_hal->create_wpa_ctrl_skt(link_sock);
                                     }
                                 }
diff --git a/agent/src/beerocks/slave/beerocks_agent_main.cpp b/agent/src/beerocks/slave/beerocks_agent_main.cpp
index 519866e04..449892fef 100644
--- a/agent/src/beerocks/slave/beerocks_agent_main.cpp
+++ b/agent/src/beerocks/slave/beerocks_agent_main.cpp
@@ -23,6 +23,10 @@ static std::shared_ptr<beerocks::nbapi::Amxrt> guarantee = nullptr;
 #include "son_slave_thread.h"
 #include "traffic_separation.h"
 
+#ifdef USE_PRPLMESH_NLHAL
+#include <bwl/nl80211_client_impl.h>
+#endif
+
 #include <bcl/beerocks_cmdu_server_factory.h>
 #include <bcl/beerocks_config_file.h>
 #include <bcl/beerocks_event_loop_impl.h>
@@ -200,27 +204,56 @@ static void fill_son_slave_config(const beerocks::config_file::sConfigSlave &bee
     // cACTION_PLATFORM_SON_SLAVE_REGISTER_RESPONSE
     agent_conf.stop_on_failure_attempts = 0;
 
+#ifdef USE_PRPLMESH_NLHAL
+    std::string fh_iface_link_idxd;
+
+    auto socket = std::make_unique<bwl::nl80211_socket>();
+        if (!socket->connect()) {
+            LOG(ERROR) << "Failed to connect nl80211 socket!";
+            return;
+        }
+
+    auto nl80211_client = std::make_unique<bwl::nl80211_client_impl>(std::move(socket));
+#endif
+
     for (const auto &iface_element : interfaces_map) {
         auto fronthaul_num    = iface_element.first;
-        auto &fronthaul_iface = iface_element.second;
-
-        // Fill configuration
-        agent_conf.radios[fronthaul_iface].enable_repeater_mode =
-            beerocks_slave_conf.enable_repeater_mode[fronthaul_num] == "1";
-        agent_conf.radios[fronthaul_iface].hostap_iface_type =
-            beerocks::utils::get_iface_type_from_string(
-                beerocks_slave_conf.hostap_iface_type[fronthaul_num]);
-        agent_conf.radios[fronthaul_iface].hostap_ant_gain =
-            (!beerocks_slave_conf.hostap_ant_gain[fronthaul_num].empty())
-                ? beerocks::string_utils::stoi(beerocks_slave_conf.hostap_ant_gain[fronthaul_num])
-                : 0;
-        agent_conf.radios[fronthaul_iface].backhaul_wireless_iface =
-            get_sta_iface_from_hostap_iface(fronthaul_iface);
-        agent_conf.radios[fronthaul_iface].backhaul_wireless_iface_filter_low =
-            (!beerocks_slave_conf.sta_iface_filter_low[fronthaul_num].empty())
-                ? beerocks::string_utils::stoi(
-                      beerocks_slave_conf.sta_iface_filter_low[fronthaul_num])
-                : 0;
+
+#ifdef USE_PRPLMESH_NLHAL
+/* Use interface name with _link<idx> as radio identifiers, previously used interface name as radio identifier,
+ * that will introduce duplicate in interface name for being used as radio identifier
+ * when there is only one 3 link mld (or) 1 slo/legacy with 1 two link mld */
+
+        std::string fh_iface_rad_idxd = iface_element.second + "##" + std::to_string(fronthaul_num);
+
+/* Fetch link idx via nl call as because idx can be any value for a band and not fixed and assigned by driver */
+        if (!nl80211_client->get_iface_with_link_idx(fh_iface_rad_idxd, fh_iface_link_idxd)) {
+            LOG(ERROR) << "Failed to base interface name for " << fh_iface_rad_idxd;
+            return;
+        }
+        LOG(INFO) << "Base iface link indexed:"<< fh_iface_link_idxd;
+        auto &fronthaul_iface = fh_iface_link_idxd;
+#else
+	auto &fronthaul_iface = iface_element.second;
+
+#endif
+            // Fill configuration
+            agent_conf.radios[fronthaul_iface].enable_repeater_mode =
+                beerocks_slave_conf.enable_repeater_mode[fronthaul_num] == "1";
+            agent_conf.radios[fronthaul_iface].hostap_iface_type =
+                beerocks::utils::get_iface_type_from_string(
+                    beerocks_slave_conf.hostap_iface_type[fronthaul_num]);
+            agent_conf.radios[fronthaul_iface].hostap_ant_gain =
+                (!beerocks_slave_conf.hostap_ant_gain[fronthaul_num].empty())
+                    ? beerocks::string_utils::stoi(beerocks_slave_conf.hostap_ant_gain[fronthaul_num])
+                    : 0;
+            agent_conf.radios[fronthaul_iface].backhaul_wireless_iface =
+                get_sta_iface_from_hostap_iface(iface_element.second);
+            agent_conf.radios[fronthaul_iface].backhaul_wireless_iface_filter_low =
+                (!beerocks_slave_conf.sta_iface_filter_low[fronthaul_num].empty())
+                    ? beerocks::string_utils::stoi(
+                        beerocks_slave_conf.sta_iface_filter_low[fronthaul_num])
+                    : 0;
     }
 
     auto db = beerocks::AgentDB::get();
@@ -619,7 +652,9 @@ int main(int argc, char *argv[])
         } else if (beerocks::net::network_utils::linux_iface_exists(interfaces[i].ifname)) {
             // if interface is not mandatory and exists
             interfaces_map[interfaces[i].radio_num] = std::string(interfaces[i].ifname);
-        }
+        } else if (0/* To check if iface exists and interface present in hostapd_phy00.<radionum>.conf*/) {
+		// then assign ifname to interfaces_map[interfaces[i].radio_num]
+	}
     }
 
     if (interfaces_map.empty()) {
diff --git a/common/beerocks/bcl/include/bcl/beerocks_defines.h b/common/beerocks/bcl/include/bcl/beerocks_defines.h
index c773bbb12..f425704ba 100644
--- a/common/beerocks/bcl/include/bcl/beerocks_defines.h
+++ b/common/beerocks/bcl/include/bcl/beerocks_defines.h
@@ -47,6 +47,12 @@ constexpr char CONF_FILES_WRITABLE_PATH[] = "./";
 // clang-format on
 #endif
 
+#define IFACE_LINK_SUFFIX "_link"
+#define IFACE_LINK_SUFFIX_LEN 5
+#define RADIOIDX_2G 0
+#define RADIOIDX_5G 1
+#define RADIOIDX_6G 2
+
 namespace beerocks {
 namespace ieee1905_1_consts {
 static constexpr int DISCOVERY_NOTIFICATION_TIMEOUT_SEC = 60;
diff --git a/common/beerocks/bwl/common/base_wlan_hal.cpp b/common/beerocks/bwl/common/base_wlan_hal.cpp
index b137789cf..0e1869eb7 100644
--- a/common/beerocks/bwl/common/base_wlan_hal.cpp
+++ b/common/beerocks/bwl/common/base_wlan_hal.cpp
@@ -256,4 +256,34 @@ bool base_wlan_hal::is_filtered_event(const std::string &opcode)
     return m_filtered_events.find(opcode) != m_filtered_events.end();
 }
 
+void base_wlan_hal::extractLinkBaseInterface(const std::string &interface_name, std::string &base_interface) {
+    size_t pos = interface_name.find("_link");
+    if (pos != std::string::npos) {
+        base_interface = interface_name.substr(0, pos);
+    } else {
+        size_t pos1 = interface_name.find("##");
+        if (pos1 != std::string::npos) {
+            base_interface = interface_name.substr(0, pos1);
+        } else {
+            base_interface = interface_name; // fallback if "_link" not found
+        }
+   }
+}
+
+void base_wlan_hal::extractLinkId(const std::string &interface_name, int &link_id) {
+    link_id = -1;
+    size_t pos_link = interface_name.find("_link");
+    if (pos_link != std::string::npos && pos_link + 5 < interface_name.length()) {
+        link_id = std::stoi(interface_name.substr(pos_link + 5));
+    }
+}
+    
+void base_wlan_hal::extractRadioId(const std::string &interface_name, int &radio_id) {
+    radio_id = -1;
+    size_t pos_hash = interface_name.find("##");
+    if (pos_hash != std::string::npos && pos_hash + 2 < interface_name.length()) {
+ 	radio_id = std::stoi(interface_name.substr(pos_hash + 2));
+    }
+}
+
 } // namespace bwl
diff --git a/common/beerocks/bwl/include/bwl/base_wlan_hal.h b/common/beerocks/bwl/include/bwl/base_wlan_hal.h
index a43c63362..bd1b672fa 100644
--- a/common/beerocks/bwl/include/bwl/base_wlan_hal.h
+++ b/common/beerocks/bwl/include/bwl/base_wlan_hal.h
@@ -155,6 +155,12 @@ public:
      */
     bool is_filtered_event(const std::string &opcode);
 
+    static void extractLinkBaseInterface(const std::string &interface_name, std::string &base_interface);
+
+    static void extractLinkId(const std::string &interface_name, int &link_id);
+    
+    static void extractRadioId(const std::string &interface_name, int &radio_id);
+
     // Public getter methods:
 public:
     /*!
diff --git a/common/beerocks/bwl/include/bwl/mon_wlan_hal.h b/common/beerocks/bwl/include/bwl/mon_wlan_hal.h
index 27e0d1026..b871f7d67 100644
--- a/common/beerocks/bwl/include/bwl/mon_wlan_hal.h
+++ b/common/beerocks/bwl/include/bwl/mon_wlan_hal.h
@@ -137,6 +137,8 @@ public:
 
     virtual int create_wpa_ctrl_skt(std::string &sock_name) = 0;
     virtual bool remove_ctrl_skt(std::string &ifname) = 0;
+ 
+    virtual bool get_mld_interface_mac(std::string &iface_name, std::string &mac_out) = 0;
 
 };
 
diff --git a/common/beerocks/bwl/shared/netlink_socket.h b/common/beerocks/bwl/include/bwl/netlink_socket.h
similarity index 100%
rename from common/beerocks/bwl/shared/netlink_socket.h
rename to common/beerocks/bwl/include/bwl/netlink_socket.h
diff --git a/common/beerocks/bwl/include/bwl/nl80211_client.h b/common/beerocks/bwl/include/bwl/nl80211_client.h
index 106d8c96f..3a2616e43 100644
--- a/common/beerocks/bwl/include/bwl/nl80211_client.h
+++ b/common/beerocks/bwl/include/bwl/nl80211_client.h
@@ -618,6 +618,12 @@ public:
     virtual bool create_new_interface(std::string &vap_name ) =0;
 
     virtual bool get_bss_link_id(std::string &interface_name,std::string &link_mac,int8_t &link_id) = 0;
+
+    virtual bool get_iface_with_link_idx(const std::string &iface_radio_idxd, std::string &iface_link_idxd) = 0;
+
+    virtual bool getMLDLinkIfaceMacAddr(std::string &interface_name, std::string &mac_out) = 0;
+
+    virtual bool get_radio_idx_from_link_id(const std::string &interface_name, int link_id, int &radio_idx) = 0;
 };
 
 } // namespace bwl
diff --git a/common/beerocks/bwl/shared/nl80211_client_impl.h b/common/beerocks/bwl/include/bwl/nl80211_client_impl.h
similarity index 94%
rename from common/beerocks/bwl/shared/nl80211_client_impl.h
rename to common/beerocks/bwl/include/bwl/nl80211_client_impl.h
index 26162fc9a..a0a6fbdc0 100644
--- a/common/beerocks/bwl/shared/nl80211_client_impl.h
+++ b/common/beerocks/bwl/include/bwl/nl80211_client_impl.h
@@ -9,7 +9,8 @@
 #define __BWL_NL80211_CLIENT_IMPL_H__
 
 #include "bwl/nl80211_client.h"
-#include "nl80211_socket.h"
+#include "bwl/nl80211_socket.h"
+#include "bwl/base_wlan_hal.h"
 
 #include <memory>
 
@@ -191,6 +192,11 @@ public:
 	    uint16_t b_8_15:8;
     } __attribute__((packed));
 
+    virtual bool get_iface_with_link_idx(const std::string &iface_radio_idxd, std::string &iface_link_idxd) override;
+
+    virtual bool getMLDLinkIfaceMacAddr(std::string &interface_name, std::string &mac_out) override;
+
+    virtual bool get_radio_idx_from_link_id(const std::string &interface_name, int link_id, int &radio_idx) override;
 
 private:
     /**
diff --git a/common/beerocks/bwl/shared/nl80211_socket.h b/common/beerocks/bwl/include/bwl/nl80211_socket.h
similarity index 98%
rename from common/beerocks/bwl/shared/nl80211_socket.h
rename to common/beerocks/bwl/include/bwl/nl80211_socket.h
index 99260487d..04f3a2430 100644
--- a/common/beerocks/bwl/shared/nl80211_socket.h
+++ b/common/beerocks/bwl/include/bwl/nl80211_socket.h
@@ -8,7 +8,7 @@
 #ifndef __BWL_NL80211_SOCKET_H__
 #define __BWL_NL80211_SOCKET_H__
 
-#include "nl_genl_socket.h"
+#include <bwl/nl_genl_socket.h>
 
 namespace bwl {
 
diff --git a/common/beerocks/bwl/shared/nl_genl_socket.h b/common/beerocks/bwl/include/bwl/nl_genl_socket.h
similarity index 97%
rename from common/beerocks/bwl/shared/nl_genl_socket.h
rename to common/beerocks/bwl/include/bwl/nl_genl_socket.h
index 63397cd91..74723dacf 100644
--- a/common/beerocks/bwl/shared/nl_genl_socket.h
+++ b/common/beerocks/bwl/include/bwl/nl_genl_socket.h
@@ -8,7 +8,7 @@
 #ifndef __BWL_NL_GENL_SOCKET_H__
 #define __BWL_NL_GENL_SOCKET_H__
 
-#include "netlink_socket.h"
+#include <bwl/netlink_socket.h>
 
 namespace bwl {
 
diff --git a/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp b/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
index d4f557238..0ab290942 100644
--- a/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
+++ b/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
@@ -196,12 +196,40 @@ static uint8_t wpa_bw_to_beerocks_bw(const std::string &chan_width)
     return (chan_width == "80+80") ? 160 : beerocks::string_utils::stoi(chan_width);
 }
 
+int ap_wlan_hal_nl80211::getRadioIdxFromLinkId(const std::string &radio_iface_name, int link_id)
+{
+    int radio_idx = -1;
+        if (!m_nl80211_client->get_radio_idx_from_link_id(radio_iface_name, link_id, radio_idx)) {
+        LOG(ERROR) << "Unable to get radio_idx from link id";
+        return -1;
+    }
+
+    LOG(INFO) << "Radio index is: " << radio_idx;
+    return radio_idx;
+}
+
 /// @brief figures out hostapd config file name by the
 //  interface name and loads its content
 prplmesh::hostapd::Configuration ap_wlan_hal_nl80211::load_hostapd_config(const std::string &radio_iface_name,const std::string &bssid)
 {
     // TODO: PRPLMESH_NLHAL To be replaced hostapd.conf selection logic
 
+    int link_id;
+    int radio_idx = -1;
+    std::string base_interface="";
+    if (!radio_iface_name.empty()) {
+        extractLinkBaseInterface(radio_iface_name, base_interface);
+        extractLinkId(radio_iface_name, link_id);
+
+	if (link_id != -1) {
+        //deriving radio_idx here is just for sanity now and is not used really in below logics
+	    radio_idx = getRadioIdxFromLinkId(base_interface, link_id);
+	    if (radio_idx == -1) {
+		return prplmesh::hostapd::Configuration("radio_idx not found");
+	    }
+	}
+    }
+
     std::vector<std::string> hostapd_cfg_names;
     std::string path = "/var/run/";
     
@@ -244,14 +272,14 @@ prplmesh::hostapd::Configuration ap_wlan_hal_nl80211::load_hostapd_config(const
         }
 
         // check if it is the right one:
-        // we are looking for the line in the vap that declares this vap: interface=radio_iface_name
+        // we are looking for the line in the vap that declares this vap: interface=base_interface
         // we could equaly ask hostapd_conf if it has this vap, but there is no such interface
-        // to do this. it should be something like: hostapd_conf.is_vap_exists(radio_iface_name);
-        if (!radio_iface_name.empty()) {
+        // to do this. it should be something like: hostapd_conf.is_vap_exists(base_interface);
+        if (!base_interface.empty()) {
           
-            if (hostapd_conf.get_vap_value(radio_iface_name, "interface") != radio_iface_name &&
-                hostapd_conf.get_vap_value(radio_iface_name, "bss") != radio_iface_name) {
-                LOG(DEBUG) << radio_iface_name << " does not exists in " << try_fname;
+            if (hostapd_conf.get_vap_value(base_interface, "interface") != base_interface &&
+                hostapd_conf.get_vap_value(base_interface, "bss") != base_interface) {
+                LOG(DEBUG) << base_interface << " does not exists in " << try_fname;
                 continue;
             } 
 
@@ -262,7 +290,7 @@ prplmesh::hostapd::Configuration ap_wlan_hal_nl80211::load_hostapd_config(const
             std::string chanlist_str = hostapd_conf.get_head_value("chanlist");
 
             if (chanlist_str.empty()) {
-                LOG(DEBUG) << "###No chanlist found in " << try_fname;
+                LOG(DEBUG) << "No chanlist found in " << try_fname;
                 continue;
             }
 
@@ -1114,7 +1142,10 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
     const std::string &bridge_ifname)
 {
     LOG(INFO) << "Update vap credentials: " << m_radio_info.iface_name<<" freq of radio is: " << m_radio_info.frequency_band;
-    
+   
+    std::string base_interface="";
+    extractLinkBaseInterface(m_radio_info.iface_name, base_interface);
+ 
     // Load hostapd config for the radio
     prplmesh::hostapd::Configuration base_conf = load_hostapd_config(m_radio_info.iface_name,"");
     if (!base_conf) {
@@ -1181,7 +1212,7 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
             
             LOG(ERROR) << "create new vap for wildcard bssid ";
 
-            std::string new_vap_name = m_radio_info.iface_name + "_" + std::to_string(new_vap_index);
+            std::string new_vap_name = base_interface + "_" + std::to_string(new_vap_index);
 
             sMacAddr new_bssid;
 
@@ -1285,16 +1316,6 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
                        << " fronthaul: " << beerocks::string_utils::bool_str(bss_it.fronthaul)
                        << " backhaul: " << beerocks::string_utils::bool_str(bss_it.backhaul);
             
-            if (bss_it.mld_id >= 0) {
-                std::string new_bss_name = get_ifname_wrt_mld_grp(bss_it.mld_id, bss);
-                if (new_bss_name == bss) {
-                        LOG(ERROR) << bss << ": no need to change bss name for this mld grouping";
-                } else {
-                        LOG(ERROR) << bss << ": Need to change the bss name to" << new_bss_name << "for this mld grouping";
-                        conf.set_create_vap_value(vap_id, bss_key_name, new_bss_name);
-                }
-            }
-
             if (!ieee80211be_sprt.empty()) {
                     conf.set_create_vap_value(vap_id, "mld_ap","1");
              }
@@ -1342,6 +1363,32 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
             conf.set_create_vap_value(vap_id, "start_disabled", "");
 
             conf.set_create_vap_value(vap_id, "wds_sta", "1");
+
+            if (bss_it.mld_id >= 0) {
+                std::string new_bss_name = get_ifname_wrt_mld_grp(bss_it.mld_id, bss);
+                if (new_bss_name == bss) {
+                        LOG(INFO) << bss << ": No need to change bss name for this mld grouping";
+                } else {
+                        LOG(INFO) << bss << ": Need to change the bss name to" << new_bss_name << "for this mld grouping";
+                        if (conf.set_create_vap_value(vap_id, bss_key_name, new_bss_name)) {
+				LOG(INFO) << "set interface/bss pass";
+			} else {
+				LOG(ERROR) << "set interface/bss fail";
+			}
+/*			auto &new_vap_id = conf.get_vap_by_bssid(tlvf::mac_to_string(bss_it.bssid));
+			LOG(INFO) << "After setting to new vap_id: " << new_vap_id;
+			if (conf.set_create_head_value(bss_key_name, new_bss_name)) {
+				LOG(INFO) << "set " << bss_key_name << " pass";
+			} else {
+				LOG(INFO) << "set " << bss_key_name << " fail";
+			}
+*/			
+			auto new_bss = conf.get_vap_value(vap_id, bss_key_name);
+			LOG(INFO) << "After setting bss value: " << new_bss;
+
+                }
+            }
+
             // escape IV
             const auto vap_iter =
                 std::find_if(m_radio_info.available_vaps.begin(), m_radio_info.available_vaps.end(),
diff --git a/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.h b/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.h
index 7f2b33071..21e461939 100644
--- a/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.h
+++ b/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.h
@@ -136,9 +136,7 @@ public:
 
     virtual void set_sta_enable(bool enable) override;
 
-    void extractBaseInterface_link(const std::string &interface_name, std::string &base_interface);
-    void extractRadioIndex_link(const std::string &interface_name, int &link_id);
-    int get_radio_idx_from_iface(const std::string &radio_iface_name, std::string base_interface, int link_id);
+    int getRadioIdxFromLinkId(const std::string &radio_iface_name, int link_id);
     prplmesh::hostapd::Configuration load_hostapd_config(const std::string &radio_iface_name, const std::string &bssid);
 
     // Protected methods:
diff --git a/common/beerocks/bwl/nl80211/base_wlan_hal_nl80211.cpp b/common/beerocks/bwl/nl80211/base_wlan_hal_nl80211.cpp
index a4d322b89..8a5c90be9 100644
--- a/common/beerocks/bwl/nl80211/base_wlan_hal_nl80211.cpp
+++ b/common/beerocks/bwl/nl80211/base_wlan_hal_nl80211.cpp
@@ -298,7 +298,12 @@ bool base_wlan_hal_nl80211::fsm_setup()
                     }
 
                     // Store the current interface index
-                    int iface_index = if_nametoindex(get_iface_name().c_str());
+                    std::string iface_name = get_iface_name();
+                    std::string base_iface_name;
+
+                    extractLinkBaseInterface(iface_name, base_iface_name);
+                    int iface_index = if_nametoindex(base_iface_name.c_str());
+
 
                     if (iface_index <= 0) {
                         LOG(ERROR) << "Failed reading the index of interface " << get_iface_name()
@@ -306,6 +311,8 @@ bool base_wlan_hal_nl80211::fsm_setup()
 
                         return false;
                     }
+/* Todo NLHAL: All base managers will have same m_iface_index now in mlo, 
+ * need to work on assigning  unique index here */
                     m_iface_index.emplace(get_iface_name(), iface_index);
                 }
 
@@ -962,10 +969,13 @@ bool base_wlan_hal_nl80211::process_ext_events(int fd)
 std::string base_wlan_hal_nl80211::get_radio_mac()
 {
     std::string mac;
-
-    if (!beerocks::net::network_utils::linux_iface_get_mac(m_radio_info.iface_name, mac)) {
-        LOG(ERROR) << "Failed to get radio mac from ifname " << m_radio_info.iface_name;
+    nl80211_client::interface_info if_info;
+    if(m_nl80211_client->get_interface_info(m_radio_info.iface_name, if_info)){
+        mac = tlvf::mac_to_string(if_info.addr);
+    } else {
+        LOG(ERROR) << "Unable to read interface info of " << m_radio_info.iface_name;
     }
+
     LOG(DEBUG) << "radio_mac of iface " <<  m_radio_info.iface_name << "  " << mac; 
     return mac;
 }
diff --git a/common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.cpp b/common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.cpp
index 2694e6406..4d26eede2 100644
--- a/common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.cpp
+++ b/common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.cpp
@@ -322,6 +322,21 @@ bool mon_wlan_hal_nl80211::update_stations_stats(const std::string &vap_iface_na
     return true;
 }
 
+bool mon_wlan_hal_nl80211::get_mld_interface_mac(std::string &iface_name, std::string &mac_out)
+{
+    LOG(INFO) << "get_mld_interface_mac triggered for interface: " << iface_name;
+
+    if (!m_nl80211_client->getMLDLinkIfaceMacAddr(iface_name, mac_out)) {
+        LOG(ERROR) << "Unable to read interface info of " << iface_name;
+        return false;
+    }
+
+    LOG(INFO) << "MAC for interface " << iface_name << ": " << mac_out;
+
+    return true;
+}
+
+
 bool mon_wlan_hal_nl80211::update_station_qos_control_params(const std::string &vap_iface_name,
                                                              const std::string &sta_mac,
                                                              SStaQosCtrlParams &sta_qos_ctrl_params)
@@ -523,16 +538,6 @@ bool mon_wlan_hal_nl80211::process_nl80211_event(parsed_obj_map_t &parsed_obj)
     if (interface.empty()) {
         LOG(DEBUG) << "Could not find interface name.";
     }
-#if 0
-    size_t pos;
-    pos = interface.find("_link");
-
-    if(pos != std::string::npos) {
-        interface = interface.substr(0, pos);
-    }
-
-    LOG(DEBUG) << "updated interface: " << interface;
-#endif
 
     auto vap_id = get_vap_id_with_bss(interface);
     if (vap_id < 0) {
diff --git a/common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.h b/common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.h
index e0cb110eb..87a01b46a 100644
--- a/common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.h
+++ b/common/beerocks/bwl/nl80211/mon_wlan_hal_nl80211.h
@@ -79,6 +79,8 @@ public:
     virtual int create_wpa_ctrl_skt(std::string &sock_name) override;
     virtual bool remove_ctrl_skt(std::string &ifname) override;
 
+    virtual bool get_mld_interface_mac(std::string &iface_name, std::string &mac_out) override;
+
     // Protected methods:
 protected:
     virtual bool process_nl80211_event(parsed_obj_map_t &parsed_obj) override;
diff --git a/common/beerocks/bwl/shared/netlink_socket.cpp b/common/beerocks/bwl/shared/netlink_socket.cpp
index c5e6f4857..8b9be0604 100644
--- a/common/beerocks/bwl/shared/netlink_socket.cpp
+++ b/common/beerocks/bwl/shared/netlink_socket.cpp
@@ -6,7 +6,7 @@
  * See LICENSE file for more details.
  */
 
-#include "netlink_socket.h"
+#include <bwl/netlink_socket.h>
 
 #include <netlink/msg.h>
 #include <netlink/netlink.h>
diff --git a/common/beerocks/bwl/shared/nl80211_client_factory.cpp b/common/beerocks/bwl/shared/nl80211_client_factory.cpp
index 85027d8d9..0a1de8fe6 100644
--- a/common/beerocks/bwl/shared/nl80211_client_factory.cpp
+++ b/common/beerocks/bwl/shared/nl80211_client_factory.cpp
@@ -8,7 +8,7 @@
 
 #include <bwl/nl80211_client_factory.h>
 
-#include "nl80211_client_impl.h"
+#include <bwl/nl80211_client_impl.h>
 
 #include <bcl/beerocks_backport.h>
 
diff --git a/common/beerocks/bwl/shared/nl80211_client_impl.cpp b/common/beerocks/bwl/shared/nl80211_client_impl.cpp
index 5f0975f17..25b6b3171 100644
--- a/common/beerocks/bwl/shared/nl80211_client_impl.cpp
+++ b/common/beerocks/bwl/shared/nl80211_client_impl.cpp
@@ -32,7 +32,7 @@
  *
  */
 
-#include "nl80211_client_impl.h"
+#include <bwl/nl80211_client_impl.h>
 
 #include <bcl/beerocks_defines.h>
 #include <bcl/beerocks_utils.h>
@@ -52,6 +52,7 @@
 #include <math.h>
 namespace bwl {
 
+
 std::string nl80211_parse_ifname(struct nlattr *tb[])
 {
     std::string ifname;
@@ -103,21 +104,166 @@ bool nl80211_client_impl::get_interfaces(std::vector<std::string> &interfaces)
         });
 }
 
+bool nl80211_client_impl::get_iface_with_link_idx(const std::string &iface_rad_idxd, std::string &iface_link_idxd)
+{
+    int  radio_idx;
+    std::string base_interface;
+    bwl::base_wlan_hal::extractLinkBaseInterface(iface_rad_idxd, base_interface);
+    bwl::base_wlan_hal::extractRadioId(iface_rad_idxd, radio_idx);
+    int link_id = 0;
+
+    LOG(INFO) << "iface_rad_idxd: " << iface_rad_idxd << " radio:" << radio_idx <<  "base_interface: " << base_interface;
+    if (!m_socket) {
+            LOG(ERROR) << "Socket is NULL!";
+            return false;
+    }
+
+    int iface_index = if_nametoindex(base_interface.c_str());
+    if (0 == iface_index) {
+            LOG(ERROR) << "Failed to read the index of interface " << iface_rad_idxd << "base_interface:" << base_interface << ": "
+                    << strerror(errno);
+            return false;
+    }
+    LOG(INFO) << "iface_index: " << iface_index;
+
+    return m_socket.get()->send_receive_msg(
+    NL80211_CMD_GET_INTERFACE, 0,
+    [&](struct nl_msg *msg) -> bool {
+    nla_put_u32(msg, NL80211_ATTR_IFINDEX, iface_index);
+    return true;
+    },
+    [&](struct nl_msg *msg) {
+    struct nlattr *tb[NL80211_ATTR_MAX + 1];
+    struct genlmsghdr *gnlh = (struct genlmsghdr *)nlmsg_data(nlmsg_hdr(msg));
+
+    // Parse the netlink message
+    if (nla_parse(tb, NL80211_ATTR_MAX, genlmsg_attrdata(gnlh, 0), genlmsg_attrlen(gnlh, 0),
+                            NULL)) {
+    LOG(ERROR) << "Failed to parse netlink message!";
+    return;
+    }
+
+    if ((tb[NL80211_ATTR_MLO_LINKS])) {
+            LOG(INFO) << "MLO supported and MLO links attr present";
+        int ret = 0;
+        struct nlattr *attrs, *link[NL80211_ATTR_MAX + 1];
+        nla_for_each_nested(attrs, tb[NL80211_ATTR_MLO_LINKS], ret) {
+                nla_parse_nested(link, NL80211_ATTR_MAX, attrs, NULL);
+                if (link[NL80211_ATTR_WIPHY_FREQ]) {
+                    int freq = nla_get_u32(link[NL80211_ATTR_WIPHY_FREQ]);
+
+                    LOG(INFO) << "freq: " << freq ;
+                    bool match = false;
+                    switch (radio_idx) {
+                            case RADIOIDX_2G: match = (freq >= 2400 && freq <= 2500); break;
+                            case RADIOIDX_5G: match = (freq >= 4900 && freq <= 5900); break;
+                            case RADIOIDX_6G: match = (freq >= 5925 && freq <= 7125); break;
+                            default: LOG(ERROR) << "Invalid radio_type"; break;
+                    }
+
+                    if (!match) {
+                            LOG(INFO) << "Skipping MLO link with freq " << freq << " not matching radio_type: " << radio_idx;
+                            continue;
+                    }
+                    link_id = nla_get_u8(link[NL80211_ATTR_MLO_LINK_ID]);
+                    LOG(INFO) << "link_id : " << link_id;
+                }
+        }
+        iface_link_idxd = base_interface + "_link" + std::to_string(link_id);
+        LOG(INFO) << "iface_link_idxd : " << iface_link_idxd;
+        return;
+    } else {
+            iface_link_idxd = base_interface;
+            LOG(ERROR) << "iface_link_idxd : " << iface_link_idxd;
+            return;
+    }
+
+    });
+}
+
+bool nl80211_client_impl::getMLDLinkIfaceMacAddr(std::string &interface_name, std::string &mac_out) {
+    LOG(INFO) <<"INSIDE get_mld_mac_for_interface";
+
+    int  link_idx;
+    std::string base_interface;
+    bwl::base_wlan_hal::extractLinkBaseInterface(interface_name, base_interface);
+    bwl::base_wlan_hal::extractLinkId(interface_name, link_idx);
+
+    LOG(INFO) << "Interface value is " << base_interface << "radio is: " << link_idx;
+
+    if (!m_socket) {
+        LOG(ERROR) << "Socket is NULL!";
+        return false;
+    }
+
+    // Get the interface index for given interface name
+    int iface_index = if_nametoindex(base_interface.c_str());
+    if (0 == iface_index) {
+        LOG(ERROR) << "Failed to read the index of interface " << base_interface << ": "
+                   << strerror(errno);
+
+        return false;
+    }
+
+    bool mac_found = false;
+    return m_socket.get()->send_receive_msg(
+        NL80211_CMD_GET_INTERFACE, 0,
+        [&](struct nl_msg *msg) -> bool {
+            nla_put_u32(msg, NL80211_ATTR_IFINDEX, iface_index);
+
+            return true;
+        },
+        [&](struct nl_msg *msg) {
+            struct nlattr *tb[NL80211_ATTR_MAX + 1];
+            struct genlmsghdr *gnlh = (struct genlmsghdr *)nlmsg_data(nlmsg_hdr(msg));
+
+            // Parse the netlink message
+            if (nla_parse(tb, NL80211_ATTR_MAX, genlmsg_attrdata(gnlh, 0), genlmsg_attrlen(gnlh, 0),
+                          NULL)) {
+                LOG(ERROR) << "Failed to parse netlink message!";
+                return false;
+            }
+            int ret = 0;
+            struct nlattr *attrs, *link[NL80211_ATTR_MAX + 1];
+            nla_for_each_nested(attrs, tb[NL80211_ATTR_MLO_LINKS], ret) {
+                nla_parse_nested(link, NL80211_ATTR_MAX, attrs, NULL);
+                if (link[NL80211_ATTR_MLO_LINK_ID] && (nla_get_u8(link[NL80211_ATTR_MLO_LINK_ID]) == link_idx)) {
+                    if (link[NL80211_ATTR_MAC]) {
+                        const uint8_t *link_addr = static_cast<const uint8_t *>(nla_data(link[NL80211_ATTR_MAC]));
+                        mac_out = tlvf::mac_to_string(link_addr);
+                        LOG(INFO) << "Matched MAC for link index " << link_idx << ": " << mac_out;
+                        mac_found = true;
+                        return true;
+                    }
+                }
+            }
+            if (!mac_found) {
+                LOG(WARNING) << "No matching MAC found for link index " << link_idx << " on interface " << base_interface;
+            }
+            return mac_found;
+        }
+    );
+}
+
 bool nl80211_client_impl::get_interface_info(const std::string &interface_name,
                                              interface_info &interface_info)
 {
     interface_info = {};
 
+    int link_idx =0;
+    std::string base_interface;
+    bwl::base_wlan_hal::extractLinkBaseInterface(interface_name, base_interface);
+    bwl::base_wlan_hal::extractLinkId(interface_name, link_idx);
+
     if (!m_socket) {
         LOG(ERROR) << "Socket is NULL!";
         return false;
     }
 
     // Get the interface index for given interface name
-    int iface_index = if_nametoindex(interface_name.c_str());
+    int iface_index = if_nametoindex(base_interface.c_str());
     if (0 == iface_index) {
-        LOG(ERROR) << "Failed to read the index of interface " << interface_name << ": "
-                   << strerror(errno);
+        LOG(ERROR) << "Failed to read the index of interface " << interface_name << "base_interface:" << base_interface << ": "                              << strerror(errno);
 
         return false;
     }
@@ -221,54 +367,56 @@ bool nl80211_client_impl::get_interface_info(const std::string &interface_name,
                 int ret = 0;
                 struct nlattr *attrs, *link[NL80211_ATTR_MAX + 1];
                 nla_for_each_nested(attrs, tb[NL80211_ATTR_MLO_LINKS], ret) {
-                    nla_parse_nested(link, NL80211_ATTR_MAX, attrs, NULL);
-                    if (link[NL80211_ATTR_MLO_LINK_ID]) {
-                        if (link[NL80211_ATTR_MAC]) {
-                            const uint8_t *link_addr = static_cast<const uint8_t *>(nla_data(link[NL80211_ATTR_MAC]));
-
-                            tlvf::mac_from_array(link_addr, interface_info.addr);
-                        }
-                    }
-                    if (link[NL80211_ATTR_WIPHY_FREQ]) {
-                        interface_info.frequency = nla_get_u32(link[NL80211_ATTR_WIPHY_FREQ]);
-                        interface_info.channel =
-                            son::wireless_utils::freq_to_channel(interface_info.frequency);
-                        if (link[NL80211_ATTR_CHANNEL_WIDTH]) {
-
-                            // NL80211_ATTR_CENTER_FREQ1 attribute must be provided for bw 40 80 160
-                            // NL80211_ATTR_CENTER_FREQ2 attribute must be provided for bw 80P80
-                            switch (nla_get_u32(link[NL80211_ATTR_CHANNEL_WIDTH])) {
-                            case NL80211_CHAN_WIDTH_20_NOHT:
-                            case NL80211_CHAN_WIDTH_20:
-                                interface_info.bandwidth = 20;
-                                break;
-                            case NL80211_CHAN_WIDTH_40:
-                                interface_info.bandwidth = 40;
-                                break;
-                            case NL80211_CHAN_WIDTH_80:
-                                interface_info.bandwidth = 80;
-                                break;
-                            case NL80211_CHAN_WIDTH_80P80:
-                                if (link[NL80211_ATTR_CENTER_FREQ2]) {
-                                    interface_info.frequency_center2 =
-                                        nla_get_u32(link[NL80211_ATTR_CENTER_FREQ2]);
-                                }
-                                interface_info.bandwidth = 160;
-                                break;
-                            case NL80211_CHAN_WIDTH_160:
-                                interface_info.bandwidth = 160;
-                                break;
-                            default:
-                                //unsupported values: ofdm bw 5 and 10
-                                interface_info.bandwidth = 0;
-                                break;
-                            }
-                            if (interface_info.bandwidth > 20 && link[NL80211_ATTR_CENTER_FREQ1]) {
-                                interface_info.frequency_center1 =
-                                    nla_get_u32(link[NL80211_ATTR_CENTER_FREQ1]);
-                            }
-                        }
-                    }
+                nla_parse_nested(link, NL80211_ATTR_MAX, attrs, NULL);
+                if (link[NL80211_ATTR_MLO_LINK_ID] && (nla_get_u8(link[NL80211_ATTR_MLO_LINK_ID]) == link_idx)) {
+
+                               if (link[NL80211_ATTR_MAC]) {
+                                       const uint8_t *link_addr = static_cast<const uint8_t *>(nla_data(link[NL80211_ATTR_MAC]));
+
+                                       tlvf::mac_from_array(link_addr, interface_info.addr);
+                               }
+
+                               if (link[NL80211_ATTR_WIPHY_FREQ]) {
+                                       interface_info.frequency = nla_get_u32(link[NL80211_ATTR_WIPHY_FREQ]);
+                                       interface_info.channel =
+                                               son::wireless_utils::freq_to_channel(interface_info.frequency);
+                                       if (link[NL80211_ATTR_CHANNEL_WIDTH]) {
+
+                                               // NL80211_ATTR_CENTER_FREQ1 attribute must be provided for bw 40 80 160
+                                               // NL80211_ATTR_CENTER_FREQ2 attribute must be provided for bw 80P80
+                                               switch (nla_get_u32(link[NL80211_ATTR_CHANNEL_WIDTH])) {
+                                                       case NL80211_CHAN_WIDTH_20_NOHT:
+                                                       case NL80211_CHAN_WIDTH_20:
+                                                               interface_info.bandwidth = 20;
+                                                               break;
+                                                       case NL80211_CHAN_WIDTH_40:
+                                                               interface_info.bandwidth = 40;
+                                                               break;
+                                                       case NL80211_CHAN_WIDTH_80:
+                                                               interface_info.bandwidth = 80;
+                                                               break;
+                                                       case NL80211_CHAN_WIDTH_80P80:
+                                                               if (link[NL80211_ATTR_CENTER_FREQ2]) {
+                                                                       interface_info.frequency_center2 =
+                                                                               nla_get_u32(link[NL80211_ATTR_CENTER_FREQ2]);
+                                                               }
+                                                               interface_info.bandwidth = 160;
+                                                               break;
+                                                       case NL80211_CHAN_WIDTH_160:
+                                                               interface_info.bandwidth = 160;
+                                                               break;
+                                                       default:
+                                                               //unsupported values: ofdm bw 5 and 10
+                                                               interface_info.bandwidth = 0;
+                                                               break;
+                                               }
+                                               if (interface_info.bandwidth > 20 && link[NL80211_ATTR_CENTER_FREQ1]) {
+                                                       interface_info.frequency_center1 =
+                                                               nla_get_u32(link[NL80211_ATTR_CENTER_FREQ1]);
+                                               }
+                                       }
+                               }
+                       }
                 }
             }
         });
@@ -290,10 +438,13 @@ bool nl80211_client_impl::get_radio_info(const std::string &interface_name, radi
         return false;
     }
 
+    std::string base_interface;
+    bwl::base_wlan_hal::extractLinkBaseInterface(interface_name, base_interface);
+
     // Get the interface index for given interface name
-    int iface_index = if_nametoindex(interface_name.c_str());
+    int iface_index = if_nametoindex(base_interface.c_str());
     if (0 == iface_index) {
-        LOG(ERROR) << "Failed to read the index of interface " << interface_name << ": "
+        LOG(ERROR) << "Failed to read the index of interface " << interface_name << " base_interface: " << base_interface << ": "
                    << strerror(errno);
 
         return false;
@@ -972,10 +1123,13 @@ bool nl80211_client_impl::get_survey_info(const std::string &interface_name,
         return false;
     }
 
+    std::string base_interface;
+    bwl::base_wlan_hal::extractLinkBaseInterface(interface_name, base_interface);
+
     // Get the interface index for given interface name
-    int iface_index = if_nametoindex(interface_name.c_str());
+    int iface_index = if_nametoindex(base_interface.c_str());
     if (0 == iface_index) {
-        LOG(ERROR) << "Failed to read the index of interface " << interface_name << ": "
+        LOG(ERROR) << "Failed to read the index of interface " << interface_name << "base_interface:" << base_interface << ": "
                    << strerror(errno);
 
         return false;
@@ -1083,10 +1237,13 @@ bool nl80211_client_impl::set_tx_power_limit(const std::string &interface_name,
         return false;
     }
 
+    std::string base_interface;
+    bwl::base_wlan_hal::extractLinkBaseInterface(interface_name, base_interface);
+
     // Get the interface index for given interface name
-    int iface_index = if_nametoindex(interface_name.c_str());
+    int iface_index = if_nametoindex(base_interface.c_str());
     if (0 == iface_index) {
-        LOG(ERROR) << "Failed to read the index of interface '" << interface_name
+        LOG(ERROR) << "Failed to read the index of interface '" << interface_name << " base_interface: " << base_interface 
                    << "': " << strerror(errno);
 
         return false;
@@ -1110,10 +1267,13 @@ bool nl80211_client_impl::get_tx_power_dbm(const std::string &interface_name, ui
         return false;
     }
 
+    std::string base_interface;
+    bwl::base_wlan_hal::extractLinkBaseInterface(interface_name, base_interface);
+
     // Get the interface index for given interface name
-    int iface_index = if_nametoindex(interface_name.c_str());
+    int iface_index = if_nametoindex(base_interface.c_str());
     if (0 == iface_index) {
-        LOG(ERROR) << "Failed to read the index of interface '" << interface_name
+        LOG(ERROR) << "Failed to read the index of interface '" << interface_name << " base_interface: " << base_interface
                    << "': " << strerror(errno);
 
         return false;
@@ -1172,11 +1332,14 @@ bool nl80211_client_impl::channel_scan_abort(const std::string &interface_name)
         LOG(ERROR) << "interface_name is empty!";
         return false;
     }
+    
+    std::string base_interface;
+    bwl::base_wlan_hal::extractLinkBaseInterface(interface_name, base_interface);
 
     // Get the interface index for given interface name
-    int iface_index = if_nametoindex(interface_name.c_str());
+    int iface_index = if_nametoindex(base_interface.c_str());
     if (0 == iface_index) {
-        LOG(ERROR) << "Failed to read the index of interface '" << interface_name
+        LOG(ERROR) << "Failed to read the index of interface '" << interface_name << "base_interface: " << base_interface
                    << "': " << strerror(errno);
 
         return false;
@@ -1202,10 +1365,13 @@ bool nl80211_client_impl::add_key(const std::string &interface_name, const sKeyI
         return false;
     }
 
+    std::string base_interface;
+    bwl::base_wlan_hal::extractLinkBaseInterface(interface_name, base_interface);
+
     // Get the interface index for given interface name
-    int iface_index = if_nametoindex(interface_name.c_str());
+    int iface_index = if_nametoindex(base_interface.c_str());
     if (0 == iface_index) {
-        LOG(ERROR) << "Failed to read the index of interface " << interface_name << ": "
+        LOG(ERROR) << "Failed to read the index of interface " << interface_name << " base_interface: " << base_interface << ": "
                    << strerror(errno);
 
         return false;
@@ -1247,11 +1413,14 @@ bool nl80211_client_impl::add_station(const std::string &interface_name, const s
         LOG(ERROR) << "Socket is NULL!";
         return false;
     }
+    
+    std::string base_interface;
+    bwl::base_wlan_hal::extractLinkBaseInterface(interface_name, base_interface);
 
     // Get the interface index for given interface name
-    int iface_index = if_nametoindex(interface_name.c_str());
+    int iface_index = if_nametoindex(base_interface.c_str());
     if (0 == iface_index) {
-        LOG(ERROR) << "Failed to read the index of interface " << interface_name << ": "
+        LOG(ERROR) << "Failed to read the index of interface " << interface_name << " base_interface: " << base_interface << ": " 
                    << strerror(errno);
 
         return false;
@@ -1285,10 +1454,13 @@ bool nl80211_client_impl::get_key(const std::string &interface_name, sKeyInfo &k
         return false;
     }
 
+    std::string base_interface;
+    bwl::base_wlan_hal::extractLinkBaseInterface(interface_name, base_interface);
+
     // Get the interface index for given interface name
-    int iface_index = if_nametoindex(interface_name.c_str());
+    int iface_index = if_nametoindex(base_interface.c_str());
     if (0 == iface_index) {
-        LOG(ERROR) << "Failed to read the index of interface " << interface_name << ": "
+        LOG(ERROR) << "Failed to read the index of interface " << interface_name << " base_interface: " << base_interface << ": "
                    << strerror(errno);
 
         return false;
@@ -1360,10 +1532,13 @@ bool nl80211_client_impl::send_delba(const std::string &interface_name, const sM
         return false;
     }
 
+    std::string base_interface;
+    bwl::base_wlan_hal::extractLinkBaseInterface(interface_name, base_interface);
+
     // Get the interface index for given interface name
-    int iface_index = if_nametoindex(interface_name.c_str());
+    int iface_index = if_nametoindex(base_interface.c_str());
     if (0 == iface_index) {
-        LOG(ERROR) << "Failed to read the index of interface " << interface_name << ": "
+        LOG(ERROR) << "Failed to read the index of interface " << interface_name << " base_interface: " << base_interface << ": "
                    << strerror(errno);
 
         return false;
@@ -1618,6 +1793,62 @@ bool nl80211_client_impl::get_bss_link_id(std::string &interface_name,std::strin
         return true;
 }
 
+bool nl80211_client_impl::get_radio_idx_from_link_id(const std::string &interface_name, int link_id, int &radio_idx)
+{
+    int iface_index = if_nametoindex(interface_name.c_str());
+    if (iface_index == 0) {
+        LOG(ERROR) << "Failed to get interface index for " << interface_name;
+        return false;
+    }
+
+    radio_idx = -1;
+    bool success =  m_socket.get()->send_receive_msg(
+        NL80211_CMD_GET_INTERFACE, 0,
+        [&](struct nl_msg *msg) -> bool {
+            nla_put_u32(msg, NL80211_ATTR_IFINDEX, iface_index);
+
+            return true;
+        },
+       [&](struct nl_msg *msg) {
+            struct nlattr *tb[NL80211_ATTR_MAX + 1];
+            struct genlmsghdr *gnlh = (struct genlmsghdr *)nlmsg_data(nlmsg_hdr(msg));
 
+            // Parse the netlink message
+            if (nla_parse(tb, NL80211_ATTR_MAX, genlmsg_attrdata(gnlh, 0), genlmsg_attrlen(gnlh, 0),
+                          NULL)) {
+                LOG(ERROR) << "Failed to parse netlink message!";
+                return false;
+            }
+
+            int ret = 0;
+            struct nlattr *attrs, *link[NL80211_ATTR_MAX + 1];
+            nla_for_each_nested(attrs, tb[NL80211_ATTR_MLO_LINKS], ret) {
+                nla_parse_nested(link, NL80211_ATTR_MAX, attrs, NULL);
+
+                if (link[NL80211_ATTR_MLO_LINK_ID] && link[NL80211_ATTR_WIPHY_FREQ]) {
+                    int current_link_id = nla_get_u8(link[NL80211_ATTR_MLO_LINK_ID]);
+                    if (current_link_id == link_id) {
+                        int freq = nla_get_u32(link[NL80211_ATTR_WIPHY_FREQ]);
+                        LOG(INFO) <<"FREQUNCY IS: "<< freq;
+                        if (freq >= 2400 && freq <= 2500) {
+                            radio_idx = RADIOIDX_2G;
+                        } else if (freq >= 4900 && freq <= 5900) {
+                            radio_idx = RADIOIDX_5G;
+                        } else if (freq >= 5925 && freq <= 7125) {
+                            radio_idx = RADIOIDX_6G;
+                        } else {
+                                LOG(ERROR) << "Unknown frequency band: " << freq;
+                        }
+                        return true; //return once match is found
+                    }
+                }
+            }
+            if (radio_idx == -1 || !success) {
+                LOG(WARNING) << "Failed to get radio index for link ID " << link_id;
+            }
+            return false;
+        });
+    return true;
+}
 
 } // namespace bwl
diff --git a/common/beerocks/bwl/shared/nl80211_socket.cpp b/common/beerocks/bwl/shared/nl80211_socket.cpp
index 0f534582e..5c9b21f22 100644
--- a/common/beerocks/bwl/shared/nl80211_socket.cpp
+++ b/common/beerocks/bwl/shared/nl80211_socket.cpp
@@ -6,7 +6,7 @@
  * See LICENSE file for more details.
  */
 
-#include "nl80211_socket.h"
+#include <bwl/nl80211_socket.h>
 
 #include <netlink/genl/ctrl.h>
 #include <netlink/genl/family.h>
diff --git a/common/beerocks/bwl/shared/nl_genl_socket.cpp b/common/beerocks/bwl/shared/nl_genl_socket.cpp
index 51a16ff7b..5ccb56aad 100644
--- a/common/beerocks/bwl/shared/nl_genl_socket.cpp
+++ b/common/beerocks/bwl/shared/nl_genl_socket.cpp
@@ -6,7 +6,7 @@
  * See LICENSE file for more details.
  */
 
-#include "nl_genl_socket.h"
+#include <bwl/nl_genl_socket.h>
 
 #include <linux/netlink.h>
 
diff --git a/framework/platform/bpl/cfg/uci/bpl_cfg_wifi.cpp b/framework/platform/bpl/cfg/uci/bpl_cfg_wifi.cpp
index 86bf5416c..b58e61152 100644
--- a/framework/platform/bpl/cfg/uci/bpl_cfg_wifi.cpp
+++ b/framework/platform/bpl/cfg/uci/bpl_cfg_wifi.cpp
@@ -566,6 +566,8 @@ int cfg_get_sta_iface(const char iface[BPL_IFNAME_LEN], char sta_iface[BPL_IFNAM
         return RETURN_ERR;
     }
 
+    LOG(INFO) << "Section name: " << section_name;
+
     if (section_name.empty()) {
         LOG(ERROR) << "Configuration for interface " << iface << " not found";
         return RETURN_ERR;
-- 
2.34.1

